File: 2021/deno/day-6.ts
const input = await Deno.readTextFile("./input/day-6.in");

const school = input.split(",").map(Number);

const simulate = (target: number) => {
  let byDay = Array.from({ length: 9 }, () => 0);

  school.forEach((fish) => {
    byDay[fish] += 1;
  });

  let day = 0;

  while (day < target) {
    let atZero = byDay.shift();
    // keep ts happy
    if (atZero == null) throw new Error("Impossible...");

    byDay.push(atZero);
    byDay[6] = byDay[6] + atZero;

    day = day + 1;
  }

  return byDay.reduce((acc, curr) => acc + curr, 0);
};

/**
 * Part One
 */
console.log("Part One:", simulate(80));

/**
 * Part Two
 */
console.log("Part Two:", simulate(256));

File: 2021/swift/day-6.swift
import Foundation

let filename = "../input/day-6.in"

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let system = contents.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: ",").map { Int($0)! }

        let runSimulation: ([Int], Int) -> [Int] = {
            system, days in
            var simulation = system.reduce(Array(repeating: 0, count: 9)) {
                acc, curr in

                var next = acc
                next[curr] += 1
                return next
            }

            for _ in 1 ... days {
                let ripe = simulation.remove(at: 0)

                simulation[6] += ripe

                simulation.append(ripe)
            }

            return simulation
        }

        print("Part one: ", runSimulation(system, 80).reduce(0, +))
        print("Part two: ", runSimulation(system, 256).reduce(0, +))
    } catch {
        print("Error reading: ", filename, error)
    }
}

main()
